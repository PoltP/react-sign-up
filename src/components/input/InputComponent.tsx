
import React from 'react';

import { Block } from './block.styled';
import { Input } from './input.styled';
import { Label } from './label.styled';
import { ErrorText } from '../elements/error.styled';
import { IError } from '../types/form-fields';

interface IInputComponentProps {
  name: string;
  label: string;
  type: string;
  value: string;
  error?: IError;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export const InputComponent = ({ name, label, type, value, error, onChange}: IInputComponentProps) => (
  <Block>
    <Label isError={!!error}>{label}
      <Input required type={type} name={name} placeholder={label} value={value} isError={!!error} onChange={onChange} />
      <ErrorText visible={!!error}>{error?.message}</ErrorText>
    </Label>
  </Block>
);

