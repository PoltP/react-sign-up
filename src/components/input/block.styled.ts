import styled from 'styled-components';

import { mediaQueries } from '../media-queries';

export const Block = styled.div`
  width: 100%;
  display: flex;
  flex: 1;
  flex-direction: column;
  align-items: flex-start;

  margin-bottom: 30px;
  @media ${mediaQueries.laptop1440} {
     margin-bottom: 25px;
  }
  @media ${mediaQueries.tablet768} {
    margin-bottom: 15px;
  }
  @media ${mediaQueries.mobile414} {
    margin-bottom: 10px;
  }
`