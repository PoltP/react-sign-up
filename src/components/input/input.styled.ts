import styled from 'styled-components';

import { DEFAULT_THEME } from '../../theming/default-theme';
import { ITheme } from '../../theming/ITheme';

export const Input = styled.input<{ theme?: ITheme; isError?: boolean }>`
  background: transparent;
  outline: none;
  height: auto;
  width: 100%;

  ${({ theme, isError }) => `
    border: 2px solid ${theme.color.border};
    padding: ${theme.padding.inner};
    margin-top: ${theme.margin.inner};
    color: ${isError ? theme.color.error : theme.color.subText};
    font-family: inherit;
    font-weight: ${theme.fontWeight.regular};
    font-size: ${theme.fontSize.label};
    line-height: 120%;

    &::placeholder {
      ${isError ? `color: ${theme.color.error}` : ''};
    }

    &:-webkit-autofill::first-line {
      font-size: ${theme.fontSize.label};
    }
    &:-webkit-autofill {
      -webkit-text-fill-color: ${isError ? theme.color.error : theme.color.subText} !important;
    }
  `}
`;

Input.defaultProps = {
  theme: DEFAULT_THEME,
};
