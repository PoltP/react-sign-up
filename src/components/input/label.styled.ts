import styled from 'styled-components';

import { DEFAULT_THEME } from '../theming/default-theme';
import { ITheme } from '../theming/ITheme';

export const Label = styled.label<{ theme?: ITheme; isError?: boolean }>`
  width: 100%;
  padding-bottom: 10px;
  position: relative;

  ${({ theme, isError }) => `
    color: ${isError ? theme.color.error : theme.color.text};
    font-family: inherit;
    font-weight: ${theme.fontWeight.bold};
    font-size: ${theme.fontSize.label};
    line-height: 120%;
  `}
`

Label.defaultProps = {
  theme: DEFAULT_THEME
};