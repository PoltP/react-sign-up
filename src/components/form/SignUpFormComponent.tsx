import React, { useEffect, useState } from 'react';
import { signUp, check } from '../../services/user';
import { Container } from './container.styled';
import { Button } from './button.styled';
import { SignUpForm } from './signup-form.styled';
import { ErrorText } from '../elements/error.styled';
import { CircleSpinner } from '../elements/circle-spinner.styled';
import { InputComponent } from '../input/InputComponent';
import { i18nCreator } from '../../localization/i18n';
import { signUpValidator } from '../validation/sign-up-validator';
import { getOtherErrors, hasCritical, isServerError } from '../validation/errors';
import type { Errors, KnownField } from '../../types/form-fields';

const i18nSignUp = i18nCreator('signup');

export const SignUpFormComponent = () => {
  const [isSent, setIsSent] = useState(false);
  const [isSigningUp, setIsSigningUp] = useState(false);
  const [isCheckingUsername, setIsCheckingUsername] = useState(false);

  const [username, setUsername] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [passwordConfirm, setPasswordConfirm] = useState('');

  const [clientErrors, setClientErrors] = useState<Errors>({});
  const [serverErrors, setServerErrors] = useState<Errors>({});

  const clear = () => {
    setUsername('');
    setEmail('');
    setPassword('');
    setPasswordConfirm('');

    setClientErrors({});
    setServerErrors({});
  };

  const handleValidate = (field: KnownField) => {
    const validationErrors = signUpValidator.validateField({username, email, password, passwordConfirm}, field);
    if (validationErrors) {
      return setClientErrors({
        ...clientErrors,
        ...validationErrors
      });
    }
  };

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const field: KnownField = event.target.name as KnownField;
    const value = event.target.value;
    switch (field) {
      case 'username':
        return setUsername(value);
      case 'email':
        return setEmail(value);
      case 'password':
        return setPassword(value);
      case 'passwordConfirm':
        return setPasswordConfirm(value);
    }
  };

  const handleSubmit = (e: any) => {
    e.preventDefault();

    const user = { username, email, password, passwordConfirm };
    const validationErrors: Errors = signUpValidator.validate(user);
    setClientErrors(validationErrors);
    if (hasCritical(validationErrors) || hasCritical(serverErrors)) {
      return;
    }

    setIsSigningUp(true);
    const sendData = async () => {
      const result = await signUp(user);
      if (isServerError(result.status)) {
        setServerErrors({
          server5xx: { message: `${i18nSignUp('error')}${result.statusText}` },
        });
        return setIsSigningUp(false);
      }
      result
        .json()
        .then((data) => {
          if (data.errors) {
            setServerErrors(data.errors);
          } else {
            clear();
            setIsSent(true);
          }
        })
        .catch((e) => {
          setServerErrors({
            unknown: { message: e.message },
          });
        })
        .finally(() => setIsSigningUp(false));
    };

    sendData();
  };

  useEffect(() => {
    const checkUsername = async () => {
      setIsCheckingUsername(true);
      const result = await check(username);
      result
        .json()
        .then((data) => {
          if (data.errors) {
            setServerErrors((prevErrors) => ({
              ...prevErrors,
              ...data.errors,
            }));
          } else {
            setServerErrors((prevErrors) => ({
              ...prevErrors,
              username: undefined,
            }));
          }
        })
        .catch((e) => {
          setServerErrors((prevErrors) => ({
            ...prevErrors,
            unknown: { message: e.message },
          }));
        })
        .finally(() => setIsCheckingUsername(false));
    };

    username && checkUsername();
  }, [username]);

  const otherErrors = getOtherErrors(serverErrors);
  return (
    <Container>
      {i18nSignUp('title')}
      <SignUpForm>
        {isSent ? (
          <>{i18nSignUp('success')}</>
        ) : (
          <>
            <InputComponent
              label={i18nSignUp('email')}
              type="email"
              name="email"
              value={email}
              error={clientErrors['email'] || serverErrors['email']}
              onChange={handleChange}
              onValidate={handleValidate}
            />
            <InputComponent
              label={i18nSignUp('username')}
              type="text"
              name="username"
              value={username}
              error={clientErrors['username'] || serverErrors['username']}
              isLoading={isCheckingUsername}
              onChange={handleChange}
              onValidate={handleValidate}
            />
            <InputComponent
              label={i18nSignUp('password')}
              type="password"
              name="password"
              value={password}
              error={clientErrors['password'] || serverErrors['password']}
              onChange={handleChange}
              onValidate={handleValidate}
            />
            <InputComponent
              label={i18nSignUp('passwordConfirm')}
              type="password"
              name="passwordConfirm"
              value={passwordConfirm}
              error={clientErrors['passwordConfirm']}
              onChange={handleChange}
              onValidate={handleValidate}
            />
            <Button disabled={isSigningUp} onClick={isSigningUp ? undefined : handleSubmit}>
              {i18nSignUp(isSigningUp ? 'signingup' : 'submit')}
              {isSigningUp && <CircleSpinner />}
              <ErrorText visible={!!otherErrors.length}>{otherErrors.join('\n')}</ErrorText>
            </Button>
          </>
        )}
      </SignUpForm>
    </Container>
  );
};
