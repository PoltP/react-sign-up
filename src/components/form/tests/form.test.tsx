import React from 'react';
import { act, cleanup, fireEvent, render, screen, waitFor } from '@testing-library/react';
import { SignUpFormComponent } from '../SignUpFormComponent';
import { Button as SignUpButton } from '../button.styled';
import { Input } from '../../input/input.styled';
import { getStyledComponent, mockFetch } from './utils';

const setInputValue = (name: string, value: string) => {
    const input = getStyledComponent(Input, name);
    fireEvent.change(input, { target: { value } });
};

const fillForm = async () => {
    await waitFor(() => {
        setInputValue('email', 'pavel.a.poltavets@gmail.com');
        setInputValue('username', 'pavel poltavets');
        setInputValue('password', '12345678');
        setInputValue('passwordConfirm', '12345678');
    });
};

afterEach(cleanup);
beforeAll(() => jest.spyOn(window, 'fetch'));

describe('SignUpForm rendering', () => {
    beforeEach(() => (window.fetch as any).mockImplementation(mockFetch()));

    it('should render empty form', () => {
        const { asFragment } = render(<SignUpFormComponent />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render filled form', async () => {
        const { asFragment } = render(<SignUpFormComponent />);
        await fillForm();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should set username on change', async () => {
        render(<SignUpFormComponent />);
        act(() => setInputValue('username', 'Albert Einstein'));
        expect(await screen.findByRole((_, e: any) =>
           e.name === 'username' && e.value === 'Albert Einstein'
        )).toBeInTheDocument();
      })
});

describe('SignUpForm client errors validation', () => {
    beforeEach(() => (window.fetch as any).mockImplementation(mockFetch()));

    it('should show all client validation errors after Sign Up is clicked when fields are empty', () => {
        const { asFragment } = render(<SignUpFormComponent />);
        fireEvent.click(getStyledComponent(SignUpButton));
        expect(asFragment()).toMatchSnapshot();
    });

    it('should show error if incorrect email was set', async () => {
        const { asFragment } = render(<SignUpFormComponent />);
        await fillForm();
        act(() => setInputValue('email', 'pavel.a.poltavets'));
        fireEvent.click(getStyledComponent(SignUpButton));
        expect(asFragment()).toMatchSnapshot();
    });

    it('should show error if incorrect password confirmation was set', async () => {
        const { asFragment } = render(<SignUpFormComponent />);
        await fillForm();
        act(() => setInputValue('passwordConfirm', '11111111'));
        fireEvent.click(getStyledComponent(SignUpButton));
        expect(asFragment()).toMatchSnapshot();
    });

    it('should show error if incorrect password length was set', async () => {
        const { asFragment } = render(<SignUpFormComponent />);
        await fillForm();
        act(() => {
            setInputValue('password', '1234567');
            setInputValue('passwordConfirm', '1234567');
        });
        fireEvent.click(getStyledComponent(SignUpButton));
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('SignUpForm server errors validation', () => {
    it('should show error after check user name', async () => {
        (window.fetch as any).mockImplementationOnce(mockFetch({
            status: 400,
            errors: {
                username: {
                    code: 'already_taken',
                    message: 'This name is already taken'
                }
            }
        }));
        render(<SignUpFormComponent />);
        act(() => setInputValue('username', 'Bill Gates'));

        expect(await screen.findByText(/This name is already taken/)).toBeInTheDocument();
    });

    it('should show error if server returs throttled', async () => {
        (window.fetch as any).mockImplementation(mockFetch(undefined, {
            status: 429,
            errors: {
                non_field_errors: {
                    code: 'throttle',
                    message: 'You request was throttled. Plesae try again in 56 sec.'
                }
            }
        }));
        render(<SignUpFormComponent />);
        await fillForm();
        await waitFor(() => {
            fireEvent.click(getStyledComponent(SignUpButton));
        });

        expect(await screen.findByText(/You request was throttled. Plesae try again in 56 sec./)).toBeInTheDocument();
    });

    it('should show error if server returns random internal error 500-526', async () => {
        (window.fetch as any).mockImplementation(mockFetch(undefined, {
            status: 500 + Math.floor(26.0 * Math.random()),
            statusText: 'Internal Server Error Status Text'
        }));
        render(<SignUpFormComponent />);
        await fillForm();
        await waitFor(() => {
            fireEvent.click(getStyledComponent(SignUpButton));
        });

        expect(await screen.findByText(/Server Error: Internal Server Error Status Text/)).toBeInTheDocument();
    });
});