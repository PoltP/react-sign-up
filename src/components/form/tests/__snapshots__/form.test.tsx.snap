// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm client errors validation should show all client validation errors after Sign Up is clicked when fields are empty 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Email
          <input
            class="sc-gKAaRy coEsgZ"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value=""
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            Provide a correct email address
          </div>
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Choose username
          <input
            class="sc-gKAaRy coEsgZ"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value=""
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            Provide a valid user name
          </div>
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Password
          <input
            class="sc-gKAaRy coEsgZ"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value=""
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            Password must have at least 8 characters
          </div>
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Repeat your password
          <input
            class="sc-gKAaRy eFgpxs"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value=""
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SignUpForm client errors validation should show error if incorrect email was set 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Email
          <input
            class="sc-gKAaRy coEsgZ"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="pavel.a.poltavets"
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            Provide a correct email address
          </div>
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Choose username
          <input
            class="sc-gKAaRy eFgpxs"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value="pavel poltavets"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Password
          <input
            class="sc-gKAaRy eFgpxs"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Repeat your password
          <input
            class="sc-gKAaRy eFgpxs"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SignUpForm client errors validation should show error if incorrect password confirmation was set 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Email
          <input
            class="sc-gKAaRy eFgpxs"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="pavel.a.poltavets@gmail.com"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Choose username
          <input
            class="sc-gKAaRy eFgpxs"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value="pavel poltavets"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Password
          <input
            class="sc-gKAaRy eFgpxs"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Repeat your password
          <input
            class="sc-gKAaRy coEsgZ"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value="11111111"
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            Password confirmation doesn't match
          </div>
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SignUpForm client errors validation should show error if incorrect password length was set 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Email
          <input
            class="sc-gKAaRy eFgpxs"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="pavel.a.poltavets@gmail.com"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Choose username
          <input
            class="sc-gKAaRy eFgpxs"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value="pavel poltavets"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Password
          <input
            class="sc-gKAaRy coEsgZ"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value="1234567"
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            Password must have at least 8 characters
          </div>
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Repeat your password
          <input
            class="sc-gKAaRy eFgpxs"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value="1234567"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SignUpForm client errors validation should show error if incorrect username was set 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Email
          <input
            class="sc-gKAaRy eFgpxs"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="pavel.a.poltavets@gmail.com"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Choose username
          <input
            class="sc-gKAaRy coEsgZ"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value="pavel><poltavets"
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            Provide a valid user name
          </div>
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Password
          <input
            class="sc-gKAaRy eFgpxs"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Repeat your password
          <input
            class="sc-gKAaRy eFgpxs"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SignUpForm rendering should render empty form 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Email
          <input
            class="sc-gKAaRy eFgpxs"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value=""
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Choose username
          <input
            class="sc-gKAaRy eFgpxs"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value=""
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Password
          <input
            class="sc-gKAaRy eFgpxs"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value=""
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Repeat your password
          <input
            class="sc-gKAaRy eFgpxs"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value=""
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SignUpForm rendering should render filled form 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Email
          <input
            class="sc-gKAaRy eFgpxs"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="pavel.a.poltavets@gmail.com"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Choose username
          <input
            class="sc-gKAaRy eFgpxs"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value="pavel poltavets"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Password
          <input
            class="sc-gKAaRy eFgpxs"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Repeat your password
          <input
            class="sc-gKAaRy eFgpxs"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SignUpForm server errors validation should forbid signing up when there is an error after checking user name 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Email
          <input
            class="sc-gKAaRy eFgpxs"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="pavel.a.poltavets@gmail.com"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Choose username
          <input
            class="sc-gKAaRy coEsgZ"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value="pavel poltavets"
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            This name is already taken
          </div>
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Password
          <input
            class="sc-gKAaRy eFgpxs"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Repeat your password
          <input
            class="sc-gKAaRy eFgpxs"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SignUpForm server errors validation should hide all client validation errors after Sign Up is clicked when fields are set but username has server error 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM bUacdj"
  >
    Sign up to Coins.ph!
    <form
      class="sc-hKFxyN dbVKjX"
    >
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Email
          <input
            class="sc-gKAaRy eFgpxs"
            name="email"
            placeholder="Email"
            required=""
            type="email"
            value="pavel.a.poltavets@gmail.com"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd cxNKDc"
        >
          Choose username
          <input
            class="sc-gKAaRy coEsgZ"
            name="username"
            placeholder="Choose username"
            required=""
            type="text"
            value="pavel poltavets"
          />
          <div
            class="sc-gtsrHT cFYBYZ"
          >
            This name is already taken
          </div>
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Password
          <input
            class="sc-gKAaRy eFgpxs"
            name="password"
            placeholder="Password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-jSFjdj ejmATF"
      >
        <label
          class="sc-iCoGMd exJjzr"
        >
          Repeat your password
          <input
            class="sc-gKAaRy eFgpxs"
            name="passwordConfirm"
            placeholder="Repeat your password"
            required=""
            type="password"
            value="12345678"
          />
          <div
            class="sc-gtsrHT dElqeF"
          />
        </label>
      </div>
      <div
        class="sc-dlnjwi kAkWwW"
      >
        Sign up
        <div
          class="sc-gtsrHT dElqeF"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
